### PR TITLE
Skip nil-value keys to make metadata reproducible

### PR DESCRIPTION
**Pull request by nobu**
## What was the end-user or developer problem that led to this PR?

Nil-value keys in a mapping end with a space or not depending on libyaml versions, and result metadata are different per platforms.

## What is your fix for the problem, implemented in this PR?

This commit makes to skip such keys to make metadata reproducible accross platforms.

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

File: lib/rubygems/psych_tree.rb [hunk 1]
```diff
@@ -14,6 +14,10 @@ def visit_String(str)
         @emitter.scalar str, nil, nil, false, true, quote
       end
 
+      def visit_Hash(o)
+        super(o.dup.delete_if {|_, v| v.nil?})
```
[the remark this answers is not in the document]
Comment (nobu) [reply lib/rubygems/psych_tree.rb:18]: `Hash#reject` returns new `Hash` always, not the same class of the receiver.

[the remark this answers is not in the document]
Comment (nobu) [reply lib/rubygems/psych_tree.rb]: Hmmm, you mean that `Gem::Specification` should not use subclasses of `Hash`?

**Comment by segiddins** [comment]
I tried a different approach that I haven't made a PR for yet, but I'm a bit wary of dropping _all_ nil values in all hashes in the serialized gemspec.

```ruby
  def to_yaml(opts = {}) # :nodoc:
    Gem.load_yaml

    # Because the user can switch the YAML engine behind our
    # back, we have to check again here to make sure that our
    # psych code was properly loaded, and load it if not.
    unless Gem.const_defined?(:NoAliasYAMLTree)
      require_relative "psych_tree"
    end

    builder = Gem::NoAliasYAMLTree.create
    builder << self
    ast = builder.tree

    require "stringio"
    io = StringIO.new
    io.set_encoding Encoding::UTF_8

    Psych::Visitors::Emitter.new(io).accept(ast)

    yaml = io.string
    # normalize yaml output to make gem builds more reproducible
    yaml.gsub!(" !!null \n", " \n")
    yaml.gsub!(": \n", ":\n") # work around libyaml bug fixed in 0.2.5, which isn't always installed by default
    yaml
  end
```

root cause is https://github.com/ruby/psych/issues/535#issuecomment-1063890389

File: lib/rubygems/psych_tree.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinemde) [reply lib/rubygems/psych_tree.rb]: Interesting. I didn't realize there was a subclass of Hash involved. We can leave this. Thanks for explaining.

**Comment by nobu** [comment]
> I tried a different approach that I haven't made a PR for yet, but I'm a bit wary of dropping _all_ nil values in all hashes in the serialized gemspec.
>
> ```ruby
>     yaml.gsub!(": \n", ":\n") # work around libyaml bug fixed in 0.2.5, which isn't always installed by default
> ```

I also considered this first, but not sure if there can never be a space with a colon at EOL other than this case.

**Comment by deivid-rodriguez** [comment]
What would be the risk of going with @nobu's approach? Seems neat.

**Review by segiddins — APPROVED** [review]

File: lib/rubygems/psych_tree.rb [hunk 2]
```diff
@@ -14,6 +14,10 @@ def visit_String(str)
         @emitter.scalar str, nil, nil, false, true, quote
       end
 
+      def visit_Hash(o)
+        super(o.dup.delete_if {|_, v| v.nil? })
```
Comment (segiddins) [inline lib/rubygems/psych_tree.rb:18]:
```suggestion
        super(o.compact)
```

since `Hash#compact` is available on Ruby 3+ ?

**Review by segiddins — APPROVED** [review]

**Pull request merged**